File: pykernel/tests/test_env_particle.py
"""Tests for pykernel.env_particle — Environmental Particle Extension.

Mirrors C++ test_env_particle.cpp structure.
VSEPR-SIM 3.0.1
"""

import math
import pytest

from pykernel.env_particle import (
    Vec3, EnvParticleKind, EnvParticle, BeadProps,
    PlantEnvResponse, shade_factor, kind_name,
    sun_deposition, wind_force, drying_kernel,
    interact_env_particle,
    RootLocalState, RootChaosCoeffs, root_chaos_factor,
    PolySegment, PiecewiseRootPoly, default_root_poly,
    RootGrowthLimits, root_growth_modifier,
    LeafGateCoeffs, LeafLocalState,
    leaf_generation_gate, leaf_gate_signal, leaf_expansion_rate,
    EnvEnergyTerms, accumulate_env, advance_particle,
)


# =====================================================================
# Particle construction
# =====================================================================

class TestConstruction:
    def test_default_kind(self):
        p = EnvParticle()
        assert p.kind == EnvParticleKind.Sun

    def test_wind_kind(self):
        p = EnvParticle(kind=EnvParticleKind.Wind)
        assert p.kind == EnvParticleKind.Wind

    def test_kind_name_sun(self):
        assert kind_name(EnvParticleKind.Sun) == "SUN"

    def test_kind_name_wind(self):
        assert kind_name(EnvParticleKind.Wind) == "WIND"

    def test_fields(self):
        p = EnvParticle(intensity=2.5, energy=100, lifetime=5)
        assert p.intensity == 2.5
        assert p.energy == 100
        assert p.lifetime == 5


# =====================================================================
# Sun deposition
# =====================================================================

class TestSunDeposition:
    def test_basic(self):
        bead = BeadProps(projected_area=2.0, transmissivity=0.5, photo_response=1.0)
        sun = EnvParticle(kind=EnvParticleKind.Sun, intensity=3.0)
        assert sun_deposition(bead, sun) == pytest.approx(3.0)

    def test_half_shade(self):
        bead = BeadProps(projected_area=2.0, transmissivity=0.5, photo_response=1.0)
        sun = EnvParticle(kind=EnvParticleKind.Sun, intensity=3.0)
        assert sun_deposition(bead, sun, 0.5) == pytest.approx(1.5)

    def test_full_occlusion(self):
        bead = BeadProps(projected_area=2.0, transmissivity=0.5, photo_response=1.0)
        sun = EnvParticle(kind=EnvParticleKind.Sun, intensity=3.0)
        assert sun_deposition(bead, sun, 1.0) == pytest.approx(0.0)

    def test_zero_area(self):
        bead = BeadProps(projected_area=0.0, transmissivity=0.5)
        sun = EnvParticle(kind=EnvParticleKind.Sun, intensity=3.0)
        assert sun_deposition(bead, sun) == pytest.approx(0.0)

    def test_high_photo(self):
        bead = BeadProps(projected_area=1.0, transmissivity=0.5, photo_response=3.0)
        sun = EnvParticle(kind=EnvParticleKind.Sun, intensity=3.0)
        assert sun_deposition(bead, sun) == pytest.approx(4.5)


# =====================================================================
# Wind force
# =====================================================================

class TestWindForce:
    def test_x_direction(self):
        bead = BeadProps(drag_coeff=0.5, projected_area=1.0, flexibility=1.0)
        wind = EnvParticle(kind=EnvParticleKind.Wind, velocity=Vec3(10, 0, 0))
        F = wind_force(bead, wind)
        assert F.x == pytest.approx(50.0)
        assert F.y == pytest.approx(0.0)

    def test_diagonal(self):
        bead = BeadProps(drag_coeff=0.5, projected_area=1.0, flexibility=1.0)
        wind = EnvParticle(velocity=Vec3(3, 4, 0))
        F = wind_force(bead, wind)
        assert F.norm() == pytest.approx(12.5)

    def test_direction_matches(self):
        bead = BeadProps(drag_coeff=0.5, projected_area=1.0, flexibility=1.0)
        wind = EnvParticle(velocity=Vec3(3, 4, 0))
        F = wind_force(bead, wind)
        Fn = F.norm()
        assert F.x / Fn == pytest.approx(0.6, abs=0.01)
        assert F.y / Fn == pytest.approx(0.8, abs=0.01)

    def test_zero_relative(self):
        bead = BeadProps(velocity=Vec3(10, 0, 0))
        wind = EnvParticle(velocity=Vec3(10, 0, 0))
        F = wind_force(bead, wind)
        assert F.norm() == pytest.approx(0.0)

    def test_flexibility(self):
        bead = BeadProps(drag_coeff=0.5, projected_area=1.0, flexibility=0.5)
        wind = EnvParticle(velocity=Vec3(10, 0, 0))
        F = wind_force(bead, wind)
        assert F.x == pytest.approx(25.0)


# =====================================================================
# Drying kernel
# =====================================================================

class TestDrying:
    def test_basic(self):
        bead = BeadProps(projected_area=2.0)
        wind = EnvParticle(velocity=Vec3(5, 0, 0))
        assert drying_kernel(bead, wind, 0.05) == pytest.approx(0.5)

    def test_zero_wind(self):
        bead = BeadProps(projected_area=2.0)
        wind = EnvParticle()
        assert drying_kernel(bead, wind, 0.05) == pytest.approx(0.0)

    def test_custom_coeff(self):
        bead = BeadProps(projected_area=2.0)
        wind = EnvParticle(velocity=Vec3(10, 0, 0))
        assert drying_kernel(bead, wind, 0.1) == pytest.approx(2.0)


# =====================================================================
# Unified interaction
# =====================================================================

class TestUnified:
    def test_sun_deposits(self):
        bead = BeadProps(projected_area=1.5, transmissivity=0.9)
        sun = EnvParticle(kind=EnvParticleKind.Sun, intensity=2.0)
        r = interact_env_particle(bead, sun)
        assert r.dE_sun > 0
        assert r.photo_bias == pytest.approx(r.dE_sun * 0.15)
        assert r.dF_wind.norm() == pytest.approx(0.0)

    def test_wind_forces(self):
        bead = BeadProps(projected_area=1.5, drag_coeff=0.47)
        wind = EnvParticle(kind=EnvParticleKind.Wind, velocity=Vec3(8, 0, 0))
        r = interact_env_particle(bead, wind)
        assert r.dF_wind.x > 0
        assert r.drying_rate > 0
        assert r.stress_bias > 0
        assert r.dE_sun == pytest.approx(0.0)

    def test_response_addition(self):
        a = PlantEnvResponse(dE_sun=1, dF_wind=Vec3(2, 0, 0), drying_rate=0.1)
        b = PlantEnvResponse(dE_sun=3, dF_wind=Vec3(0, 1, 0), stress_bias=0.5)
        c = a + b
        assert c.dE_sun == pytest.approx(4.0)
        assert c.dF_wind.x == pytest.approx(2.0)
        assert c.dF_wind.y == pytest.approx(1.0)


# =====================================================================
# Root chaos factor
# =====================================================================

class TestRootChaos:
    def test_default(self):
        s = RootLocalState()
        assert root_chaos_factor(s) == pytest.approx(1.05)

    def test_high_moisture(self):
        s = RootLocalState(moisture=1.0)
        assert root_chaos_factor(s) == pytest.approx(1.2)

    def test_dense_soil_clamp(self):
        s = RootLocalState(soil_density=1.0)
        assert root_chaos_factor(s) == pytest.approx(0.98)

    def test_extreme_chaos_clamp(self):
        s = RootLocalState(soil_density=0.0, chaos_perturb=50.0)
        assert root_chaos_factor(s) == pytest.approx(1.8)

    def test_custom_bounds(self):
        cc = RootChaosCoeffs(alpha0=2.5, lo=0.5, hi=3.0)
        s = RootLocalState()
        g = root_chaos_factor(s, cc)
        assert 0.5 <= g <= 3.0


# =====================================================================
# Piecewise polynomial
# =====================================================================

class TestPiecewisePoly:
    def test_dry_zone(self):
        p = default_root_poly()
        assert p.eval(0.1) == pytest.approx(0.08)

    def test_intercept(self):
        p = default_root_poly()
        assert p.eval(0.0) == pytest.approx(0.05)

    def test_optimal(self):
        p = default_root_poly()
        assert p.eval(0.5) == pytest.approx(0.8875)

    def test_saturated(self):
        p = default_root_poly()
        assert p.eval(0.9) == pytest.approx(0.885)

    def test_optimal_beats_dry(self):
        p = default_root_poly()
        assert p.eval(0.5) > p.eval(0.1)


# =====================================================================
# Root growth modifier
# =====================================================================

class TestRootGrowth:
    def test_basic_bounds(self):
        s = RootLocalState(moisture=0.5)
        g = root_growth_modifier(s)
        assert 0.05 <= g <= 2.25

    def test_damage_reduces(self):
        s = RootLocalState(moisture=0.5)
        g0 = root_growth_modifier(s)
        s.damage = 0.8
        g1 = root_growth_modifier(s)
        assert g1 < g0

    def test_compaction_reduces(self):
        s = RootLocalState(moisture=0.5)
        g0 = root_growth_modifier(s)
        s.compaction = 0.8
        g1 = root_growth_modifier(s)
        assert g1 < g0

    def test_nutrient_boost(self):
        s = RootLocalState(moisture=0.5)
        g0 = root_growth_modifier(s)
        s.nutrient_grad = 0.5
        g1 = root_growth_modifier(s)
        assert g1 > g0

    def test_low_sun_reduces(self):
        s = RootLocalState(moisture=0.5)
        g0 = root_growth_modifier(s)
        s.sun_coupling = 0.1
        g1 = root_growth_modifier(s)
        assert g1 < g0


# =====================================================================
# Leaf generation gate
# =====================================================================

class TestLeafGate:
    def test_no_resources(self):
        assert not leaf_generation_gate(LeafLocalState())

    def test_sufficient(self):
        s = LeafLocalState(energy_local=2.0, sunlight=1.5, hydration=1.0)
        assert leaf_generation_gate(s)

    def test_damage_suppresses(self):
        s = LeafLocalState(energy_local=2.0, sunlight=1.5, hydration=1.0,
                           damage_sum=3.0)
        assert not leaf_generation_gate(s)

    def test_signal_positive(self):
        s = LeafLocalState(energy_local=2.0, sunlight=1.5, hydration=1.0)
        assert leaf_gate_signal(s) > 0

    def test_expansion_positive(self):
        s = LeafLocalState(energy_local=2.0, sunlight=1.5, hydration=1.0)
        assert leaf_expansion_rate(s) > 0

    def test_expansion_zero_below(self):
        s = LeafLocalState()
        assert leaf_expansion_rate(s) == pytest.approx(0.0)


# =====================================================================
# Energy decomposition
# =====================================================================

class TestEnergy:
    def test_accumulate(self):
        bead = BeadProps(position=Vec3(1, 0, 0))
        r = PlantEnvResponse(dE_sun=5, dF_wind=Vec3(2, 0, 0),
                             drying_rate=0.3, photo_bias=0.75, stress_bias=0.16)
        E = accumulate_env([r], [bead])
        assert E.U_sun == pytest.approx(5.0)
        assert E.U_wind < 0
        assert E.U_dry == pytest.approx(0.3)
        assert E.total() != 0


# =====================================================================
# Particle advection
# =====================================================================

class TestAdvection:
    def test_alive(self):
        p = EnvParticle(lifetime=2.0, velocity=Vec3(1, 0, 0))
        assert advance_particle(p, 0.5) is True
        assert p.position.x == pytest.approx(0.5)
        assert p.lifetime == pytest.approx(1.5)

    def test_dead(self):
        p = EnvParticle(lifetime=0.5, velocity=Vec3(1, 0, 0))
        assert advance_particle(p, 1.0) is False


# =====================================================================
# Full pipeline
# =====================================================================

class TestPipeline:
    def test_full(self):
        sun = EnvParticle(kind=EnvParticleKind.Sun, intensity=2.0,
                          velocity=Vec3(0, -1, 0))
        wind = EnvParticle(kind=EnvParticleKind.Wind,
                           velocity=Vec3(5, 0, 0))

        bead = BeadProps(projected_area=1.0, transmissivity=0.7,
                         photo_response=1.2, drag_coeff=0.5,
                         flexibility=0.9)

        r_sun = interact_env_particle(bead, sun)
        r_wind = interact_env_particle(bead, wind)
        combined = r_sun + r_wind

        rs = RootLocalState(moisture=0.6, sun_coupling=r_sun.dE_sun)
        gamma = root_chaos_factor(rs)
        growth = root_growth_modifier(rs)

        ls = LeafLocalState(energy_local=r_sun.dE_sun,
                            sunlight=sun.intensity,
                            hydration=rs.moisture)
        leaf = leaf_generation_gate(ls)

        assert combined.dE_sun > 0
        assert combined.dF_wind.norm() > 0
        assert 0.98 <= gamma <= 1.8
        assert growth > 0


# =====================================================================
# Parametrized sweep: chaos factor across soil densities
# =====================================================================

_DENSITIES = [0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]


class TestChaosSweep:
    @pytest.mark.parametrize("rho", _DENSITIES)
    def test_bounded(self, rho: float):
        s = RootLocalState(soil_density=rho)
        g = root_chaos_factor(s)
        assert 0.98 <= g <= 1.80

    def test_monotonic_decrease(self):
        """Higher soil density -> lower or equal gamma."""
        vals = [root_chaos_factor(RootLocalState(soil_density=r))
                for r in _DENSITIES]
        for i in range(1, len(vals)):
            assert vals[i] <= vals[i - 1] + 1e-9

File: tests/test_env_particle.cpp
/**
 * test_env_particle.cpp  —  Environmental Particle Extension Tests
 * ================================================================
 * VSEPR-SIM 3.0.1
 *
 * 45 tests covering:
 *   T1-T5:   Particle construction and kind classification
 *   T6-T10:  Sun deposition kernel
 *   T11-T15: Wind force kernel
 *   T16-T18: Drying kernel
 *   T19-T22: Unified interaction dispatch
 *   T23-T27: Root chaos factor (Gamma_i)
 *   T28-T32: Piecewise polynomial root response
 *   T33-T37: Root growth modifier (poly * logic * chaos)
 *   T38-T42: Leaf generation gate (Heaviside)
 *   T43:     Energy decomposition accumulation
 *   T44:     Particle advection and lifetime decay
 *   T45:     Full pipeline (particle -> bead -> response -> root -> leaf)
 */

#include "atomistic/models/env_particle.hpp"
#include <cassert>
#include <cmath>
#include <cstdio>
#include <vector>

using namespace atomistic;
using namespace atomistic::environment;

static int g_pass = 0;
static int g_fail = 0;

#define CHECK(expr, msg)                                                \
    do {                                                                \
        if (expr) {                                                     \
            ++g_pass;                                                   \
            std::printf("  T%-2d PASS: %s\n", g_pass + g_fail, msg);   \
        } else {                                                        \
            ++g_fail;                                                   \
            std::printf("  T%-2d FAIL: %s  [%s:%d]\n",                  \
                        g_pass + g_fail, msg, __FILE__, __LINE__);      \
        }                                                               \
    } while (0)

static bool near(double a, double b, double tol = 1e-6) {
    return std::abs(a - b) < tol;
}

// =====================================================================
// T1-T5: Particle construction
// =====================================================================

static void test_particle_construction() {
    EnvParticle p;
    CHECK(p.kind == EnvParticleKind::Sun, "default kind is Sun");

    p.kind = EnvParticleKind::Wind;
    CHECK(p.kind == EnvParticleKind::Wind, "kind set to Wind");

    CHECK(std::string(kind_name(EnvParticleKind::Sun)) == "SUN",
          "kind_name Sun");
    CHECK(std::string(kind_name(EnvParticleKind::Wind)) == "WIND",
          "kind_name Wind");

    EnvParticle sun;
    sun.kind = EnvParticleKind::Sun;
    sun.intensity = 2.5;
    sun.energy = 100.0;
    sun.lifetime = 5.0;
    CHECK(sun.intensity == 2.5 && sun.energy == 100.0 && sun.lifetime == 5.0,
          "sun particle fields");
}

// =====================================================================
// T6-T10: Sun deposition
// =====================================================================

static void test_sun_deposition() {
    BeadProps bead;
    bead.projected_area = 2.0;
    bead.transmissivity = 0.5;
    bead.photo_response = 1.0;

    EnvParticle sun;
    sun.kind = EnvParticleKind::Sun;
    sun.intensity = 3.0;

    // dE = A * I * T * S * Phi = 2 * 3 * 0.5 * 1 * 1 = 3.0
    double dE = sun_deposition(bead, sun, 0.0);
    CHECK(near(dE, 3.0), "sun deposition basic (dE=3.0)");

    // With 50% occlusion: dE = 3.0 * 0.5 = 1.5
    dE = sun_deposition(bead, sun, 0.5);
    CHECK(near(dE, 1.5), "sun deposition with 50% shade");

    // Full occlusion
    dE = sun_deposition(bead, sun, 1.0);
    CHECK(near(dE, 0.0), "sun deposition full occlusion");

    // Zero area
    bead.projected_area = 0.0;
    dE = sun_deposition(bead, sun, 0.0);
    CHECK(near(dE, 0.0), "sun deposition zero area");

    // High photo response
    bead.projected_area = 1.0;
    bead.photo_response = 3.0;
    dE = sun_deposition(bead, sun, 0.0);
    // 1 * 3 * 0.5 * 3 * 1 = 4.5
    CHECK(near(dE, 4.5), "sun deposition high photo response");
}

// =====================================================================
// T11-T15: Wind force
// =====================================================================

static void test_wind_force() {
    BeadProps bead;
    bead.projected_area = 1.0;
    bead.drag_coeff = 0.5;
    bead.flexibility = 1.0;
    bead.velocity = {0, 0, 0};

    EnvParticle wind;
    wind.kind = EnvParticleKind::Wind;
    wind.velocity = {10, 0, 0};

    // F = C_d * A * |v_rel|^2 * Psi = 0.5 * 1 * 100 * 1 = 50
    Vec3 F = wind_force(bead, wind);
    CHECK(near(F.x, 50.0) && near(F.y, 0.0) && near(F.z, 0.0),
          "wind force +x direction (F=50)");

    // Diagonal wind
    wind.velocity = {3, 4, 0};  // |v| = 5, |v|^2 = 25
    F = wind_force(bead, wind);
    double Fmag = norm(F);
    // F_mag = 0.5 * 1 * 25 * 1 = 12.5
    CHECK(near(Fmag, 12.5), "wind force diagonal (|F|=12.5)");

    // Direction check: F should be along (3,4,0)/5
    CHECK(near(F.x / Fmag, 0.6, 0.01) && near(F.y / Fmag, 0.8, 0.01),
          "wind force direction matches velocity");

    // Zero relative velocity
    bead.velocity = {10, 0, 0};
    wind.velocity = {10, 0, 0};
    F = wind_force(bead, wind);
    CHECK(near(norm(F), 0.0), "wind force zero relative velocity");

    // Flexibility modifier
    bead.velocity = {0, 0, 0};
    wind.velocity = {10, 0, 0};
    bead.flexibility = 0.5;
    F = wind_force(bead, wind);
    // 0.5 * 1 * 100 * 0.5 = 25
    CHECK(near(F.x, 25.0), "wind force with flexibility=0.5");
}

// =====================================================================
// T16-T18: Drying kernel
// =====================================================================

static void test_drying_kernel() {
    BeadProps bead;
    bead.projected_area = 2.0;
    bead.velocity = {0, 0, 0};

    EnvParticle wind;
    wind.kind = EnvParticleKind::Wind;
    wind.velocity = {5, 0, 0};

    // drying = coeff * |v_rel| * A = 0.05 * 5 * 2 = 0.5
    double dr = drying_kernel(bead, wind, 0.05);
    CHECK(near(dr, 0.5), "drying kernel basic");

    // Zero wind
    wind.velocity = {0, 0, 0};
    dr = drying_kernel(bead, wind, 0.05);
    CHECK(near(dr, 0.0), "drying kernel zero wind");

    // Custom coefficient
    wind.velocity = {10, 0, 0};
    dr = drying_kernel(bead, wind, 0.1);
    // 0.1 * 10 * 2 = 2.0
    CHECK(near(dr, 2.0), "drying kernel custom coeff");
}

// =====================================================================
// T19-T22: Unified interaction
// =====================================================================

static void test_unified_interaction() {
    BeadProps bead;
    bead.projected_area = 1.5;
    bead.transmissivity = 0.9;
    bead.photo_response = 1.0;
    bead.drag_coeff = 0.47;
    bead.flexibility = 1.0;

    // Sun particle
    EnvParticle sun;
    sun.kind = EnvParticleKind::Sun;
    sun.intensity = 2.0;

    PlantEnvResponse r = interact_env_particle(bead, sun);
    CHECK(r.dE_sun > 0.0, "unified: sun deposits energy");
    CHECK(near(r.photo_bias, r.dE_sun * 0.15), "unified: photo_bias = 15% of dE");
    CHECK(near(norm(r.dF_wind), 0.0), "unified: sun produces no wind force");

    // Wind particle
    EnvParticle wind;
    wind.kind = EnvParticleKind::Wind;
    wind.velocity = {8, 0, 0};

    r = interact_env_particle(bead, wind);
    CHECK(r.dF_wind.x > 0.0, "unified: wind produces force");
    CHECK(r.drying_rate > 0.0, "unified: wind causes drying");
    CHECK(r.stress_bias > 0.0, "unified: wind causes stress");
    CHECK(near(r.dE_sun, 0.0), "unified: wind deposits no sun energy");
}

// =====================================================================
// T23-T27: Root chaos factor
// =====================================================================

static void test_root_chaos() {
    RootChaosCoeffs cc;
    RootLocalState s;

    // Default: Gamma = 1.0 + 0.3*0.5 + 0.15*0 + (-0.2)*0.5 + 0.1*0
    //        = 1.0 + 0.15 - 0.1 = 1.05
    double g = root_chaos_factor(s, cc);
    CHECK(near(g, 1.05), "root chaos default (Gamma=1.05)");

    // High moisture -> higher gamma
    s.moisture = 1.0;
    g = root_chaos_factor(s, cc);
    // 1.0 + 0.3 - 0.1 = 1.2
    CHECK(near(g, 1.2), "root chaos high moisture (Gamma=1.2)");

    // Dense soil -> lower gamma
    s.moisture = 0.5;
    s.soil_density = 1.0;
    g = root_chaos_factor(s, cc);
    // 1.0 + 0.15 - 0.2 = 0.95 -> clamped to 0.98
    CHECK(near(g, 0.98), "root chaos dense soil clamped to 0.98");

    // Very high stochastic -> clamped to 1.8
    s.soil_density = 0.0;
    s.chaos_perturb = 50.0;
    g = root_chaos_factor(s, cc);
    CHECK(near(g, 1.8), "root chaos extreme stochastic clamped to 1.8");

    // Custom bounds
    RootChaosCoeffs cc2;
    cc2.lo = 0.5;
    cc2.hi = 3.0;
    cc2.alpha0 = 2.5;
    s = {};
    g = root_chaos_factor(s, cc2);
    CHECK(g >= 0.5 && g <= 3.0, "root chaos custom bounds");
}

// =====================================================================
// T28-T32: Piecewise polynomial
// =====================================================================

static void test_piecewise_poly() {
    PiecewiseRootPoly p = default_root_poly();

    // Dry zone: x = 0.1 -> 2*(0.01) + 0.1*0.1 + 0.05 = 0.02 + 0.01 + 0.05 = 0.08
    double v = p.eval(0.1);
    CHECK(near(v, 0.08), "piecewise poly dry zone (x=0.1)");

    // Exact at x_dry boundary uses dry polynomial
    v = p.eval(0.0);
    CHECK(near(v, 0.05), "piecewise poly at x=0 (intercept)");

    // Optimal zone: x = 0.5
    v = p.eval(0.5);
    // -0.5*(0.125) + 1.2*(0.25) - 0.3*(0.5) + 0.8
    // = -0.0625 + 0.3 - 0.15 + 0.8 = 0.8875
    CHECK(near(v, 0.8875), "piecewise poly optimal zone (x=0.5)");

    // Saturated zone: x = 0.9
    v = p.eval(0.9);
    // -1.5*(0.81) + 2.0*(0.9) + 0.3 = -1.215 + 1.8 + 0.3 = 0.885
    CHECK(near(v, 0.885), "piecewise poly saturated zone (x=0.9)");

    // Monotonicity check: growth peaks in optimal zone
    double v_dry = p.eval(0.1);
    double v_opt = p.eval(0.5);
    CHECK(v_opt > v_dry, "poly optimal > dry growth");
}

// =====================================================================
// T33-T37: Root growth modifier
// =====================================================================

static void test_root_growth_modifier() {
    PiecewiseRootPoly poly = default_root_poly();
    RootChaosCoeffs cc;
    RootGrowthLimits lim;

    RootLocalState s;
    s.moisture = 0.5;  // optimal zone

    double g = root_growth_modifier(s, poly, cc, lim);
    CHECK(g > 0.0 && g <= lim.final_hi, "root modifier basic in bounds");

    // Damage penalty
    s.damage = 0.8;  // above limit
    double g_dmg = root_growth_modifier(s, poly, cc, lim);
    CHECK(g_dmg < g, "root modifier damage reduces growth");

    // Compaction penalty
    s.damage = 0.0;
    s.compaction = 0.8;
    double g_comp = root_growth_modifier(s, poly, cc, lim);
    CHECK(g_comp < g, "root modifier compaction reduces growth");

    // Nutrient boost
    s.compaction = 0.0;
    s.nutrient_grad = 0.5;
    double g_nut = root_growth_modifier(s, poly, cc, lim);
    CHECK(g_nut > g, "root modifier nutrient boost");

    // Low sun penalty
    s.nutrient_grad = 0.0;
    s.sun_coupling = 0.1;
    double g_sun = root_growth_modifier(s, poly, cc, lim);
    CHECK(g_sun < g, "root modifier low sun reduces growth");
}

// =====================================================================
// T38-T42: Leaf generation gate
// =====================================================================

static void test_leaf_gate() {
    LeafGateCoeffs c;
    c.theta = 2.0;

    LeafLocalState s;
    // signal = 0 + 0 + 0 - 0 - 2 = -2 -> no leaf
    CHECK(!leaf_generation_gate(s, c), "leaf gate: no resources -> no leaf");

    // Provide enough resources
    s.energy_local = 2.0;
    s.sunlight = 1.5;
    s.hydration = 1.0;
    // signal = 2 + 1.2 + 0.6 - 0 - 2 = 1.8 -> leaf
    CHECK(leaf_generation_gate(s, c), "leaf gate: sufficient resources -> leaf");

    // Damage kills it
    s.damage_sum = 3.0;
    // signal = 2 + 1.2 + 0.6 - 4.5 - 2 = -2.7 -> no leaf
    CHECK(!leaf_generation_gate(s, c), "leaf gate: damage suppresses");

    // Signal value is continuous
    s.damage_sum = 0.0;
    double sig = leaf_gate_signal(s, c);
    CHECK(sig > 0.0, "leaf gate signal positive when gate open");

    // Expansion rate
    double rate = leaf_expansion_rate(s, c, 0.1);
    CHECK(rate > 0.0, "leaf expansion rate positive when gate open");

    // No expansion below threshold
    s.energy_local = 0.0;
    s.sunlight = 0.0;
    s.hydration = 0.0;
    rate = leaf_expansion_rate(s, c, 0.1);
    CHECK(near(rate, 0.0), "leaf expansion zero below threshold");
}

// =====================================================================
// T43: Energy decomposition
// =====================================================================

static void test_energy_decomposition() {
    BeadProps bead;
    bead.projected_area = 1.0;
    bead.position = {1, 0, 0};

    PlantEnvResponse r;
    r.dE_sun = 5.0;
    r.dF_wind = {2, 0, 0};
    r.drying_rate = 0.3;
    r.photo_bias = 0.75;
    r.stress_bias = 0.16;

    std::vector<PlantEnvResponse> resps = {r};
    std::vector<BeadProps> beads = {bead};

    EnvEnergyTerms E = accumulate_env(resps, beads);
    CHECK(near(E.U_sun, 5.0), "energy: U_sun accumulated");
    CHECK(E.U_wind < 0.0, "energy: U_wind = -F.r < 0 for +x force/position");
    CHECK(near(E.U_dry, 0.3), "energy: U_dry");
    CHECK(E.total() != 0.0, "energy: nonzero total");
}

// =====================================================================
// T44: Particle advection
// =====================================================================

static void test_particle_advection() {
    EnvParticle p;
    p.position = {0, 0, 0};
    p.velocity = {1, 0, 0};
    p.lifetime = 2.0;
    p.omega = 0.0;

    bool alive = advance_particle(p, 0.5);
    CHECK(alive, "particle alive after partial dt");
    CHECK(near(p.position.x, 0.5), "particle advected to x=0.5");
    CHECK(near(p.lifetime, 1.5), "particle lifetime decreased");

    // Kill it
    alive = advance_particle(p, 2.0);
    CHECK(!alive, "particle dead after lifetime exceeded");
}

// =====================================================================
// T45: Full pipeline
// =====================================================================

static void test_full_pipeline() {
    // 1. Create sun and wind particles
    EnvParticle sun;
    sun.kind = EnvParticleKind::Sun;
    sun.intensity = 2.0;
    sun.velocity = {0, -1, 0};

    EnvParticle wind;
    wind.kind = EnvParticleKind::Wind;
    wind.velocity = {5, 0, 0};

    // 2. Create bead
    BeadProps bead;
    bead.projected_area = 1.0;
    bead.transmissivity = 0.7;
    bead.photo_response = 1.2;
    bead.drag_coeff = 0.5;
    bead.flexibility = 0.9;

    // 3. Interact
    PlantEnvResponse r_sun = interact_env_particle(bead, sun);
    PlantEnvResponse r_wind = interact_env_particle(bead, wind);
    PlantEnvResponse combined = r_sun + r_wind;

    // 4. Root chaos
    RootLocalState rs;
    rs.moisture = 0.6;
    rs.sun_coupling = r_sun.dE_sun;
    double gamma = root_chaos_factor(rs);

    // 5. Root growth
    PiecewiseRootPoly poly = default_root_poly();
    double growth = root_growth_modifier(rs, poly);

    // 6. Leaf gate
    LeafLocalState ls;
    ls.energy_local = r_sun.dE_sun;
    ls.sunlight = sun.intensity;
    ls.hydration = rs.moisture;
    bool leaf = leaf_generation_gate(ls);

    CHECK(combined.dE_sun > 0.0 &&
          norm(combined.dF_wind) > 0.0 &&
          gamma >= 0.98 && gamma <= 1.8 &&
          growth > 0.0 &&
          (leaf || !leaf),
          "full pipeline (sun+wind->bead->root+leaf)");
}

// =====================================================================
// Main
// =====================================================================

int main() {
    std::printf("=== Environmental Particle Extension Tests (45) ===\n\n");

    test_particle_construction();  // T1-T5
    test_sun_deposition();         // T6-T10
    test_wind_force();             // T11-T15
    test_drying_kernel();          // T16-T18
    test_unified_interaction();    // T19-T22
    test_root_chaos();             // T23-T27
    test_piecewise_poly();         // T28-T32
    test_root_growth_modifier();   // T33-T37
    test_leaf_gate();              // T38-T42 (actually T38-T43 with 6 checks)
    test_energy_decomposition();   // T43 -> T44-T47
    test_particle_advection();     // T44 -> T48-T50
    test_full_pipeline();          // T45 -> T51

    std::printf("\n");
    if (g_fail == 0) {
        std::printf("=== ALL %d TESTS PASSED ===\n", g_pass);
    } else {
        std::printf("=== %d PASSED, %d FAILED ===\n", g_pass, g_fail);
    }
    return g_fail;
}
